### ISSUE TITLE
Bloom Filter Epic

### ISSUE THREAD

**Issue by denesb** [scylladb/scylladb#17972]
Bloom filters allow for quick and cheap checks for whether a partition exists in an sstable or not, without doing any I/O.
Bloom filters are allowed to provide false-positive answers (check says that the partition is in the sstable, but its not), but it can never provide false-negative answers. They are created based on two parameters: partition count (often an estimate) and require false-positive chance (probability of false-positive results). To satisfy these two requirements, a filter of a certain size is required. The filter size is proportional to the partition count and inversely proportional to the false-positive chance.

Too large or too small filters are problematic. Too small filters will result in too many false-positives, increasing read latency and read memory consumption (extra I/O). Too large filters consume a lot of extra memory (filters are always kept in memory).
Mis-sized filters are a result of bad partition estimates: our estimate of how many partitions are in an sstable is either an over or under estimate. In the extreme case, either severe under or over estimates can cause OOM. Under estimates because a lot of sstables are opened for reads (this is rare) and over estimates because the filters end up using up all memory of the shard.

This issue is an epic, covering the effort of resolving all our current issues with bloom filters.
List of issues currently part of this epic:
[ ] https://github.com/scylladb/scylladb/issues/17747
[ ] https://github.com/scylladb/scylladb/issues/2024

Additional work (no issues opened/found yet):
* Improve estimates in streaming/repair/las
* Use better overlap estimation for compaction
* Take estimated expiry percentage into account when estimating partition count
* Store exact partition count in scylla-metadata component (currently we estimate based on the summary)

**Comment by mykaul** [comment]
I'm not sure (well, I'm quite sure it should not be) it fits /this/ epic for bloom filter, but where can we put an action item to research and explore alternatives, such as Ribbon (see https://arxiv.org/pdf/2103.02515.pdf and https://rocksdb.org/blog/2021/12/29/ribbon-filter.html ), and/or https://github.com/RoaringBitmap/CRoaring or https://cs-people.bu.edu/zczhu/files/SHaMBa-VLDB-PhD-Workshop.pdf ?

**Comment by raphaelsc** [comment]
Btw, with tablets, estimation in load-and-stream will be a lot better after it was made tablet aware, so we estimate in each sstable only the range that belongs to a given tablet. we know that for each tablet, that all partitions (from sstables being loaded) in that tablet range will land in all of the replicas.

**Comment by raphaelsc** [comment]
> I'm not sure (well, I'm quite sure it should not be) it fits /this/ epic for bloom filter, but where can we put an action item to research and explore alternatives, such as Ribbon (see https://arxiv.org/pdf/2103.02515.pdf and https://rocksdb.org/blog/2021/12/29/ribbon-filter.html ), and/or https://github.com/RoaringBitmap/CRoaring or https://cs-people.bu.edu/zczhu/files/SHaMBa-VLDB-PhD-Workshop.pdf ?

also I'd like to recommend that we take a look at compressed filters https://www.eecs.harvard.edu/~michaelm/NEWWORK/postscripts/cbf2.pdf

**Comment by denesb** [comment]
Won't changing the filter format/compress it require us to come up with our own file format?

**Comment by mykaul** [comment]
> Won't changing the filter format/compress it require us to come up with our own file format?

I would like to believe we can adopt a different file format, not invent our own. As an example, there are many projects using Roaring - I hope there's some format that can be reused there. But yes, it will not be what we use today, I reckon.

**Comment by denesb** [comment]
> > Won't changing the filter format/compress it require us to come up with our own file format?
> 
> I would like to believe we can adopt a different file format, not invent our own. As an example, there are many projects using Roaring - I hope there's some format that can be reused there. But yes, it will not be what we use today, I reckon.

By "our own file format" I mean a new version code, which will not clash with those introduced by C* (past present or future). It can be almost the same as `ME`, with the only difference being the filter format. But this would already throw away compatibility, which we went to great pains to keep so far. That said, once the incompatibility genie is out of the bottle, we might want to do much more than just change the filter. So I think this will be an epic of its own.

**Comment by mykaul** [comment]
> > > Won't changing the filter format/compress it require us to come up with our own file format?
> > 
> > 
> > I would like to believe we can adopt a different file format, not invent our own. As an example, there are many projects using Roaring - I hope there's some format that can be reused there. But yes, it will not be what we use today, I reckon.
> 
> By "our own file format" I mean a new version code, which will not clash with those introduced by C* (past present or future). It can be almost the same as `ME`, with the only difference being the filter format. But this would already throw away compatibility, which we went to great pains to keep so far. That said, once the incompatibility genie is out of the bottle, we might want to do much more than just change the filter. So I think this will be an epic of its own.

Yes to all the above, but Q: when we import sstables from elsewhere - don't we re-create the bloom filter anyway?

**Comment by denesb** [comment]
> Yes to all the above, but Q: when we import sstables from elsewhere - don't we re-create the bloom filter anyway?

When using -LaS -- yes. When using plain `refresh` -- no.

**Comment by mykaul** [comment]
> > Yes to all the above, but Q: when we import sstables from elsewhere - don't we re-create the bloom filter anyway?
> 
> When using -LaS -- yes. When using plain `refresh` -- no.

(I assume the latter will go away in the future, when we'll have to 'tabletize' (C) imported sstables anyway?)

**Comment by avikivity** [comment]
For load-and-stream, and streaming/repair in general, we can not generate the bloom filter in the first place, and always reshape. Reshape will know the partition count and generate a correct bloom filter.

Since we usually reshape, we don't lose much by forcing it.

**Comment by mykaul** [comment]
Ref https://github.com/scylladb/scylladb/issues/2130

**Comment by raphaelsc** [comment]
> For load-and-stream, and streaming/repair in general, we can not generate the bloom filter in the first place, and always reshape. Reshape will know the partition count and generate a correct bloom filter.
> 
> Since we usually reshape, we don't lose much by forcing it.

Yes, that's essentially what we were doing before that fix for repair estimation, and it was working apparently :-) SSTables are usually disjoint, and until reshape happens, we're adding only 1 sstable to the "sstable count per read".

**Comment by raphaelsc** [comment]
> Yes, the fix exposed a bug that was much worse that what it fixed.

Maybe we should revert it until we have Lakshmi's fix?

**Comment by denesb** [comment]
So looks like the fix is to estimate 0 partitions for streaming and repair? These sstables cover a narrow range, so the sstable set should be able to do a good enough job of filtering them, and reshape should restore the filters.

**Comment by raphaelsc** [comment]
> So looks like the fix is to estimate 0 partitions for streaming and repair? These sstables cover a narrow range, so the sstable set should be able to do a good enough job of filtering them, and reshape should restore the filters.

I think so. there would be 1 sstable in maintenance set overlapping with a given token. as the patch that wired estimation is causing much more harm than good, it might be a good idea to revert it. we can re-revert it once we know estimations are near-perfection. but today we know they aren't.

**Comment by raphaelsc** [comment]
this is the commit https://github.com/scylladb/scylladb/commit/4592bbe18207e9b25d485ba37103de9a35574a1d

**Comment by avikivity** [comment]
Estimating 0 or dropping bloom filters aren't a replacement for getting a good estimation.

Estimating 0 means we'll have a high amount of access to those sstables. Yes, they'll get filtered by partitioned_sstable_set, but 1 or 2 will survive.

Dropping bloom filters is a big hammer. We might drop the wrong ones. We won't OOM but performance will deteriorate.

Having said all that, I don't actually know how to estimate partition count for repair. The only thing that comes to mind is to estimate based on repair history, but we don't keep that data.

**Comment by denesb** [comment]
I think we can do something like I proposed in https://github.com/scylladb/scylladb/pull/18141#issuecomment-2031679125.
When repair is running as part of RBNO, we can use our best estimate of how much data the remote node has. When we run a genuine repair, we can use a fraction of this (maybe a configurable), like 0.1.

**Comment by avikivity** [comment]
I think 0.1 of the data is a good estimate. If it's too high, the harm isn't great. If it's too low, the bloom filter will be useless, but at least we tried.

**Comment by mykaul** [comment]
Unsure what left here for 6.2.

**Comment by denesb** [comment]
https://github.com/scylladb/scylladb/issues/2024 is the only item left todo here. We will do it in Q4.

**Comment by mykaul** [comment]
All but https://github.com/scylladb/scylladb/issues/2024 - which is less important, is done. Moving this epic to 6.4.

**Comment by denesb** [comment]
All sub-items are done/closed. Closing as complete.
